### DISCUSSION TITLE
The road to Piccolo version 1.0

### DISCUSSION THREAD

**Discussion by dantownsend** [piccolo-orm/piccolo#326] [General]
Early next year I'd like to release Piccolo version 1.0.

I've been using Piccolo in production for several years, but I know that version 1.0 is a very important signifier, and I really want to nail the user experience before then. Piccolo API and Piccolo Admin will also go to version 1.0 at the same time.

It will be a great opportunity to spread the word about Piccolo.

If there are any real annoyances anyone has with Piccolo let me know so we can fix them before 1.0.

**Comment by sinisaos** [comment]
@dantownsend These are my few thoughts.

**Piccolo ORM** already has a lot of good features for working with the database. Simple and easy-to-understand syntax is Piccolo's main strength. For some edge cases we can always write raw SQL within Piccolo ORM. The main missing point that most users want to have are `M2M` relationship (discussed #44 and #98).

**Piccolo Admin** has all the key features (crud, filtering, sorting, CSV export, etc.) and more (automatic form generation, table configuration etc.) and it is already much more than other admin interfaces. A really nice feature that all users want to have is the ability to upload files with a specific data record (discussed [here](https://github.com/piccolo-orm/piccolo_admin/issues/109)).

**Piccolo API** has all the key features and making a web API is really easy because it already offers everything out of the box (security, crud, ratelimiting etc). Only thing I think is missing is the **optional** cursor pagination (which doesn't have to work with `Piccolo Admin` and `FastAPIWrapper`) for large datasets (PR [here](https://github.com/piccolo-orm/piccolo_api/pull/94) and discussed [here](https://github.com/piccolo-orm/piccolo_api/issues/15)).

**Reply by dantownsend** [reply]
@sinisaos Thanks, I think you've summarised things well.

Like you say, there aren't too many big things which need adding.

**Reply by dantownsend** [reply]
``M2M`` is done now - so one step closer to v1.

The next thing I'd like to add is reverse foreign key lookups. It would work very similarly to M2M.

**Comment by dantownsend** [comment]
We're getting closer to version 1.

We now have a custom documentation theme, which will give Piccolo a nice lick of paint for v1:

https://github.com/piccolo-orm/piccolo_theme
